dapp/request_limitation/count.go

```diff
@@ -0,0 +1,37 @@
+package request_limitation
+
+import "sync"
+
+// a simple increase and decrease
+// request limitation.
+type Count struct {
+	lock              sync.Mutex
+	count             uint
+	max               uint
+	canNotIncreaseErr error
+}
+
+func NewCount(max uint, canNotIncreaseErr error) *Count {
+	return &Count{
+		lock:              sync.Mutex{},
+		max:               max,
+		count:             0,
+		canNotIncreaseErr: canNotIncreaseErr,
+	}
+}
+
+func (c *Count) Increase() error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.count+1 > c.max {
+		return c.canNotIncreaseErr
+	}
+	c.count++
+	return nil
+}
+
+func (c *Count) Decrease() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.count--
```

Comment:
Might want to put some check in place to not allow counter to decrease if count is 0, otherwise a scenario like below can eventually happen

```
package main

import (
	"fmt"
)


func main() {
	var count uint
	count = 0
	count--
	fmt.Println(count) // 4294967295
}
```

Reply:
Fixed.